fine_tunes.go

```diff
@@ -0,0 +1,279 @@
+	if err := getResponseObject(resp, &output); err != nil {
+		return nil, err
+	}
+	return &output, nil
+}
+
+// FineTune retrieves a fine-tuning job from the user's organization.
+//
+// See: https://beta.openai.com/docs/api-reference/fine-tunes/retrieve
+func (c *client) FineTune(ctx context.Context, fineTuneID string) (*FineTuneObject, error) {
+	req, err := c.newRequest(ctx, "GET", fmt.Sprintf("/fine-tunes/%s", fineTuneID), nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := c.performRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	output := FineTuneObject{}
+	if err := getResponseObject(resp, &output); err != nil {
+		return nil, err
+	}
+	return &output, nil
+}
+
+// CancelFineTune cancels a fine-tuning job from the user's organization.
+//
+// See: https://beta.openai.com/docs/api-reference/fine-tunes/cancel
+func (c *client) CancelFineTune(ctx context.Context, fineTuneID string) (*FineTuneObject, error) {
+	req, err := c.newRequest(ctx, "POST", fmt.Sprintf("/fine-tunes/%s/cancel", fineTuneID), nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := c.performRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	output := FineTuneObject{}
+	if err := getResponseObject(resp, &output); err != nil {
+		return nil, err
+	}
+	return &output, nil
+}
+
+// FineTuneEvents lists the events that belong to a fine-tuning job.
+//
+// See: https://beta.openai.com/docs/api-reference/fine-tunes/events
+func (c *client) FineTuneEvents(ctx context.Context, request FineTuneEventsRequest) (*FineTuneEventsResponse, error) {
+	request.Stream = false
+	req, err := c.newRequest(ctx, "GET", fmt.Sprintf("/fine-tunes/%s/events", request.FineTuneID), nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := c.performRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	output := FineTuneEventsResponse{}
+	if err := getResponseObject(resp, &output); err != nil {
+		return nil, err
+	}
+	return &output, nil
+}
+
+// FineTuneStreamEvents streams the events that belong to a fine-tuning job.
+//
+// See: https://beta.openai.com/docs/api-reference/fine-tunes/events
+func (c *client) FineTuneStreamEvents(ctx context.Context, request FineTuneEventsRequest, onData func(*FineTuneEvent)) error {
+	request.Stream = true
+	req, err := c.newRequest(ctx, "GET", fmt.Sprintf("/fine-tunes/%s/events", request.FineTuneID), nil)
+	if err != nil {
+		return err
+	}
+	resp, err := c.performRequest(req)
+	if err != nil {
+		return err
+	}
+
+	reader := bufio.NewReader(resp.Body)
+	defer resp.Body.Close()
+
+	for {
+		line, err := reader.ReadBytes('\n')
+		if err != nil {
+			return err
+		}
+
+		line = bytes.TrimSpace(line)
+		if !bytes.HasPrefix(line, dataPrefix) {
+			continue
+		}
+		line = bytes.TrimPrefix(line, dataPrefix)
+
+		if bytes.HasPrefix(line, streamTerminationPrefix) {
+			break
+		}
+		output := new(FineTuneEvent)
+		if err := json.Unmarshal(line, output); err != nil {
+			return fmt.Errorf("invalid json stream data: %v", err)
```

Comment:
it could be nice to share the code of the streaming of fine tunes with completion streaming. just becuase I noticed some of the handling is slightly different and this error output is different. that being said maybe it isn't worth the effort for only two repetitions of the code so far.

Reply:
Yeah I completely agree and would love to have a shared set of methods for them. I am sure OpenAI is just going to keep adding more streaming APIs, hopefully the ChatGPT API?